### DISCUSSION TITLE
Writing tests for DoNothingBCs and potential FlowConvection kernel

### DISCUSSION THREAD

**Discussion by ghost** [shannon-lab/zapdos#107] [General / Ideas]
Hi all, I have two ideas I would like to suggest:

1. @keniley1 passed down a FlowConvection kernel to me that he made but never ended up using, so it's not featured on any current branches/forks of Zapdos but I find myself using it for my research. I would like to try my hand at writing a test for this kernel to verify that it is working properly. I've attached the source code as a single text file.
[FlowConvection_src.txt](https://github.com/shannon-lab/zapdos/files/8010381/FlowConvection_src.txt)

2. As @csdechant described in my discussion post about the DoNothing BCs (https://github.com/shannon-lab/zapdos/discussions/106#discussion-3855666), there seems to be a bug that requires the user to enter material property values even when "use_material_props = true". I would also like to try my hand at removing this bug, but this is of less priority that idea (1) above. 

Let me know what y'all think!

@gsgall The FlowConvection kernel issue may be of interest to you!

**Comment by cticenhour** [comment]
If you find that you need `FlowConvection` in your regular work within Zapdos, it should definitely be included and tested. As a bit of guidance on what the test should include, I tend to prefer the smallest test that can verify the capability mathematically for a first test of new code. We aim for a test to run for only a couple of seconds (ideally). Some physics-based cases can't do that, but the simplest test of the mathematical implementation of a new code object generally can.

As far as (2), I was a bit late on jumping in to that discussion, but I think it's an excellent way to learn more about how we initialize member variables from input file parameters in MOOSE and C++ (and how to catch when things go wrong). Feel free to ping me about that one if you have questions. I have an idea for the fix, but it's a good exercise if you want to get more development experience.

**Reply by ghost** [reply]
I don't think I have enough experience with FEM to rigorously test the mathematics of the code, The best I can probably do is come up with possible simple physics problem that might "suggest" it is working.

**Reply by csdechant** [reply]
Are you familiar with Method of Manufactured Solutions (MMS)? It is often used for the validation / verification of codes. In short, if one has a system of equation they want to test, one can define a know solution and derive the input values needed to solve for that known solution. Then the user enter the require input values into the code they are testing and compare the code's solution to the known solution. With a convergence test, this checks two things: did the developer coded the equation correctly and does the code's convergence scale correctly (Note: this does not check if one coded the correct physics for the problem, one would have to do some kind of benchmarking or experimental validation for that). When defining the know solution, make sure to:
- use physical limitations (e.g. non-negative density and energy, etc.)
- the solution should have non-zero derivatives for the highest order (e.g. if the set of equations has a second order derivative, don't use f(x) = x as a solution). Using trig function is common with MMS for this reason.

The MOOSE website has some more information at https://mooseframework.inl.gov/python/mms.html

**Reply by ghost** [reply]
Awesome, I will look into it!
